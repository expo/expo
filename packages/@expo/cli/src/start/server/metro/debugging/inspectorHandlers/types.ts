import type { unstable_Device } from '@react-native/dev-middleware';

// @ts-expect-error until we sort out an issue with private members
export type DebuggerMetadata = NonNullable<unstable_Device['_debuggerConnection']>;

export interface InspectorHandler {
  /**
   * Intercept a message coming from the device, modify or respond to it through `this._sendMessageToDevice`.
   * Return `true` if the message was handled, this will stop the message propagation.
   */
  onDeviceMessage?(message: DeviceRequest | DeviceResponse, info: DebuggerMetadata): boolean;

  /**
   * Intercept a message coming from the debugger, modify or respond to it through `socket.send`.
   * Return `true` if the message was handled, this will stop the message propagation.
   */
  onDebuggerMessage?(message: DebuggerRequest, info: DebuggerMetadata): boolean;
}

/**
 * The outline of a basic Chrome DevTools Protocol request, either from device or debugger.
 * Both the request and response parameters could be optional, use `never` to enforce these fields.
 */
export type CdpMessage<
  Method extends string = string,
  Request extends object = object,
  Response extends object = object,
> = {
  id: number;
  method: Method;
  params: Request;
  result: Response;
};

export type DeviceRequest<M extends CdpMessage = CdpMessage> = Pick<M, 'method' | 'params'>;
export type DeviceResponse<M extends CdpMessage = CdpMessage> = Pick<M, 'id' | 'result'>;

export type DebuggerRequest<M extends CdpMessage = CdpMessage> = Pick<
  M,
  'id' | 'method' | 'params'
>;
export type DebuggerResponse<M extends CdpMessage = CdpMessage> = Pick<M, 'result'>;
