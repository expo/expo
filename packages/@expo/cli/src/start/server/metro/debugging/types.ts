import type { CustomMessageHandlerConnection } from '@react-native/dev-middleware';

export type Connection = CustomMessageHandlerConnection;
export type DeviceMetadata = Connection['device'];
export type DebuggerMetadata = Connection['debugger'];
export type Page = Connection['page'];
export type TargetCapabilityFlags = Page['capabilities'];

/**
 * The outline of a basic Chrome DevTools Protocol request, either from device or debugger.
 * Both the request and response parameters could be optional, use `never` to enforce these fields.
 */
export type CdpMessage<
  Method extends string = string,
  Request extends object = object,
  Response extends object = object,
> = {
  id: number;
  method: Method;
  params: Request;
  result: Response;
};

export type DeviceRequest<M extends CdpMessage = CdpMessage> = Pick<M, 'method' | 'params'>;
export type DeviceResponse<M extends CdpMessage = CdpMessage> = Pick<M, 'id' | 'result'>;

export type DebuggerRequest<M extends CdpMessage = CdpMessage> = Pick<
  M,
  'id' | 'method' | 'params'
>;
export type DebuggerResponse<M extends CdpMessage = CdpMessage> = Pick<M, 'result'>;
