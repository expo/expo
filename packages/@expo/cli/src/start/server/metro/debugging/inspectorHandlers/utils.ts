import type WS from 'ws';

import { DebuggerResponse, DeviceResponse } from './types';

/**
 * Helper function to respond to a message from the debugger or device.
 * The return value is used to stop the message propagation, "canceling" further handling.
 *
 * @example ```
 *  return respond<DeviceResponse<CDP>>(socket, { id: message.id, result: {} });
 * ```
 */
export function respond<T = DeviceResponse | DebuggerResponse>(socket: WS, message: T) {
  socket.send(JSON.stringify(message));
  return true;
}

/** Known compatible debuggers that require specific workarounds */
export type DebuggerType = 'chrome' | 'vscode' | 'unknown';

// Patterns to test against user agents
const CHROME_USER_AGENT = /chrome/i;
const VSCODE_USER_AGENT = /vscode/i;

/**
 * Determine the debugger type based on the known user agent.
 */
export function getDebuggerType(userAgent?: string | null): DebuggerType {
  if (userAgent && CHROME_USER_AGENT.test(userAgent)) return 'chrome';
  if (userAgent && VSCODE_USER_AGENT.test(userAgent)) return 'vscode';
  return 'unknown';
}
