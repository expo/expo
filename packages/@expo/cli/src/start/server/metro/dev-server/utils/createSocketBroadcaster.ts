import type { RawData as WebSocketRawData } from 'ws';

import type { SocketId, SocketMap } from './createSocketMap';

const debug = require('debug')('expo:metro:dev-server:broadcaster') as typeof console.log;

export function createBroadcaster(sockets: SocketMap) {
  return function broadcast(senderSocketId: SocketId | null, message: string | WebSocketRawData) {
    // Ignore if there are no connected sockets
    if (!sockets.size) return;

    for (const [socketId, socket] of sockets) {
      if (socketId === senderSocketId) continue;

      try {
        socket.send(message);
      } catch (error) {
        debug(`Failed to broadcast message to socket "${socketId}"`, error);
      }
    }
  };
}
