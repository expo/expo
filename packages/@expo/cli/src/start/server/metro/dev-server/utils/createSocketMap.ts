import type { WebSocket } from 'ws';

const debug = require('debug')('expo:metro:dev-server:socketmap') as typeof console.log;

export type SocketId = string;
export type SocketMap = Map<string, WebSocket>;

export function createSocketMap() {
  const map: SocketMap = new Map();
  const createId = createSocketIdFactory();

  const registerSocket = (socket: WebSocket) => {
    const id = createId();
    map.set(id, socket);
    return {
      id,
      terminate: () => {
        map.delete(id);
        socket.removeAllListeners();
        socket.terminate();
      },
    };
  };

  const findSocket = (id: SocketId): WebSocket | null => {
    const socket = map.get(id);
    if (!socket) debug(`No connected socket found with ID: ${id}`);
    return socket ?? null;
  };

  return { map, registerSocket, findSocket };
}

function createSocketIdFactory() {
  let nextId = 0;
  return () => `socket#${nextId++}`;
}
