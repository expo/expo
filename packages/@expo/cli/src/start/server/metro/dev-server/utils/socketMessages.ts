import type { RawData as WebSocketRawData } from 'ws';

const debug = require('debug')('expo:metro:dev-server:messages') as typeof console.log;

/** The current websocket-based communication between Metro, CLI, and client devices */
const PROTOCOL_VERSION = 2;

/**
 * Parse the incoming raw message data and return the parsed object.
 * This returns null if the protocol version did not match expected version.
 */
export function parseRawMessage<T = Record<string, any>>(
  data: WebSocketRawData,
  isBinary: boolean
): null | T {
  if (isBinary) return null;

  try {
    const { version, ...message } = JSON.parse(data.toString()) ?? {};
    if (version === PROTOCOL_VERSION) {
      return message;
    }

    debug(
      `Received message protocol version did not match supported "${PROTOCOL_VERSION}", received: ${message.version}`
    );
  } catch (error) {
    debug(`Failed to parse message: ${error}`);
  }

  return null;
}

/**
 * Serialize any of the messages to send over websockets.
 * This adds the protocol version to the message.
 */
export function serializeMessage(message: Record<string, any>): string {
  return JSON.stringify({ ...message, version: PROTOCOL_VERSION });
}
