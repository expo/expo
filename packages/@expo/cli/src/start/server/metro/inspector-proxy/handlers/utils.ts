import type WS from 'ws';

import { DebuggerResponse, DeviceResponse } from './types';

/**
 * Helper function to respond to a message from the debugger or device.
 * The return value is used to stop the message propagation, "canceling" further handling.
 *
 * @example ```
 *  return respond<DeviceResponse<CDP>>(socket, { id: message.id, result: {} });
 * ```
 */
export function respond<T = DeviceResponse | DebuggerResponse>(socket: WS, message: T) {
  socket.send(JSON.stringify(message));
  return true;
}
