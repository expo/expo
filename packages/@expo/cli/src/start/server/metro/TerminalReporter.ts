// This file represents an abstraction on the metro TerminalReporter.
// We use this abstraction to safely extend the TerminalReporter for our own custom logging.
import chalk from 'chalk';
import { Terminal } from 'metro-core';
import UpstreamTerminalReporter from 'metro/src/lib/TerminalReporter';
import util from 'util';

import { stripAnsi } from '../../../utils/ansi';
import {
  BundleDetails,
  TerminalReportableEvent,
  TerminalReporterInterface,
} from './TerminalReporter.types';

/**
 * A standard way to log a warning to the terminal. This should not be called
 * from some arbitrary Metro logic, only from the reporters. Instead of
 * calling this, add a new type of ReportableEvent instead, and implement a
 * proper handler in the reporter(s).
 */
export function logWarning(terminal: Terminal, format: string, ...args: any[]): void {
  const str = util.format(format, ...args);
  terminal.log('%s: %s', chalk.yellow('warning'), str);
}

/**
 * Similar to `logWarning`, but for messages that require the user to act.
 */
export function logError(terminal: Terminal, format: string, ...args: any[]): void {
  terminal.log(
    '%s: %s',
    chalk.red('error'),
    // Syntax errors may have colors applied for displaying code frames
    // in various places outside of where Metro is currently running.
    // If the current terminal does not support color, we'll strip the colors
    // here.
    util.format(chalk.supportsColor ? format : stripAnsi(format), ...args)
  );
}

const XTerminalReporter = UpstreamTerminalReporter as unknown as TerminalReporterInterface;

/** Extended TerminalReporter class but with proper types and extra functionality to avoid using the `_log` method directly in subclasses. */
export class TerminalReporter extends XTerminalReporter implements TerminalReporterInterface {
  /**
   * A cache of { [buildID]: BundleDetails } which can be used to
   * add more contextual logs. BundleDetails is currently only sent with `bundle_build_started`
   * so we need to cache the details in order to print the platform info with other event types.
   */
  _bundleDetails: Map<string, BundleDetails> = new Map();

  /** Keep track of how long a bundle takes to complete */
  _bundleTimers: Map<string, number> = new Map();

  _log(event: TerminalReportableEvent): void {
    switch (event.type) {
      case 'transform_cache_reset':
        return this.transformCacheReset();
      case 'dep_graph_loading':
        return this.dependencyGraphLoading(event.hasReducedPerformance);
      case 'client_log':
        if (this.shouldFilterClientLog(event)) {
          return;
        }
        break;
    }
    return super._log(event);
  }

  /** Gives subclasses an easy interface for filtering out logs. Return `true` to skip. */
  shouldFilterClientLog(event: {
    type: 'client_log';
    level: 'trace' | 'info' | 'warn' | 'log' | 'group' | 'groupCollapsed' | 'groupEnd' | 'debug';
    data: unknown[];
  }): boolean {
    return false;
  }

  /** Cache has been reset. */
  transformCacheReset(): void {}

  /** One of the first logs that will be printed. */
  dependencyGraphLoading(hasReducedPerformance: boolean): void {}

  /**
   * Custom log event representing the end of the bundling.
   *
   * @param event event object.
   * @param duration duration of the build in milliseconds.
   */
  bundleBuildEnded(event: TerminalReportableEvent, duration: number): void {}

  /**
   * This function is exclusively concerned with updating the internal state.
   * No logging or status updates should be done at this point.
   */
  _updateState(
    event: TerminalReportableEvent & { bundleDetails?: BundleDetails; buildID?: string }
  ) {
    // Append the buildID to the bundleDetails.
    if (event.bundleDetails) {
      event.bundleDetails.buildID = event.buildID;
    }

    super._updateState(event);
    switch (event.type) {
      case 'bundle_build_done':
      case 'bundle_build_failed': {
        const startTime = this._bundleTimers.get(event.buildID);
        // Observed a bug in Metro where the `bundle_build_done` is invoked twice during a static bundle
        // i.e. `expo export`.
        if (startTime == null) {
          break;
        }

        this.bundleBuildEnded(event, startTime ? Date.now() - startTime : 0);
        this._bundleTimers.delete(event.buildID);
        break;
      }
      case 'bundle_build_started':
        this._bundleDetails.set(event.buildID, event.bundleDetails);
        this._bundleTimers.set(event.buildID, Date.now());
        break;
    }
  }
}
