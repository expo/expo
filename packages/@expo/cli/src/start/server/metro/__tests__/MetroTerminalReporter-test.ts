import { stripAnsi } from '../../../../utils/ansi';
import {
  formatUsingNodeStandardLibraryError,
  isNodeStdLibraryModule,
  MetroTerminalReporter,
  stripMetroInfo,
} from '../MetroTerminalReporter';
import { BundleDetails } from '../TerminalReporter.types';

const asBundleDetails = (value: any) => value as BundleDetails;

describe('_getBundleStatusMessage', () => {
  const buildID = '1';
  const reporter = new MetroTerminalReporter('/', {
    log: jest.fn(),
    persistStatus: jest.fn(),
    status: jest.fn(),
  });
  reporter._getElapsedTime = jest.fn(() => 100);
  reporter._bundleTimers.set(buildID, 0);

  it(`should format standard progress`, () => {
    expect(
      stripAnsi(
        reporter._getBundleStatusMessage(
          {
            bundleDetails: asBundleDetails({
              entryFile: './index.js',
              platform: 'ios',
              buildID,
            }),
            ratio: 0.5,
            totalFileCount: 100,
            transformedFileCount: 50,
          },
          'in_progress'
        )
      )
    ).toMatchInlineSnapshot(`"iOS ./index.js ▓▓▓▓▓▓▓▓░░░░░░░░ 50.0% ( 50/100)"`);
  });
  it(`should format standard progress for a server invocation`, () => {
    expect(
      stripAnsi(
        reporter._getBundleStatusMessage(
          {
            bundleDetails: asBundleDetails({
              entryFile: './index.js',
              platform: 'ios',
              buildID,
              customTransformOptions: {
                environment: 'node',
              },
            }),
            ratio: 0.5,
            totalFileCount: 100,
            transformedFileCount: 50,
          },
          'in_progress'
        )
      )
    ).toMatchInlineSnapshot(`"Server ./index.js ▓▓▓▓▓▓▓▓░░░░░░░░ 50.0% ( 50/100)"`);
  });

  it(`should format standard progress at 0%`, () => {
    expect(
      stripAnsi(
        reporter._getBundleStatusMessage(
          {
            bundleDetails: asBundleDetails({
              entryFile: './index.js',
              platform: 'android',
              buildID,
            }),
            ratio: 0,
            totalFileCount: 100,
            transformedFileCount: 0,
          },
          'in_progress'
        )
      )
    ).toMatchInlineSnapshot(`"Android ./index.js ░░░░░░░░░░░░░░░░  0.0% (  0/100)"`);
  });
  it(`should format complete loading`, () => {
    expect(
      stripAnsi(
        reporter._getBundleStatusMessage(
          {
            bundleDetails: asBundleDetails({
              entryFile: './index.js',
              platform: 'android',
              buildID,
            }),
            ratio: 1.0,
            totalFileCount: 100,
            transformedFileCount: 100,
          },
          'done'
        )
      )
    ).toMatchInlineSnapshot(`"Android Bundling complete 100ms"`);
  });
  it(`should format failed loading`, () => {
    expect(
      stripAnsi(
        reporter._getBundleStatusMessage(
          {
            bundleDetails: asBundleDetails({
              entryFile: './index.js',
              platform: 'android',
              buildID,
            }),
            ratio: 1.0,
            totalFileCount: 100,
            transformedFileCount: 100,
          },
          'failed'
        )
      )
    ).toMatchInlineSnapshot(`"Android Bundling failed 100ms"`);
  });
});

describe(stripMetroInfo, () => {
  it(`sanitizes`, () => {
    const input = [
      'Unable to resolve module path from /Users/evanbacon/Documents/GitHub/examples/with-apple-auth/App.js: path could not be found within the project or in these directories:',
      '  node_modules',
      '  ../../node_modules',
      '  ../../../../node_modules',
      '',
      'If you are sure the module exists, try these steps:',
      ' 1. Clear watchman watches: watchman watch-del-all',
      ' 2. Delete node_modules and run yarn install',
      " 3. Reset Metro's cache: yarn start --reset-cache",
      ' 4. Remove the cache: rm -rf /tmp/metro-*',
      '\x1B[0m \x1B[90m 77 |\x1B[39m }\x1B[0m',
      '\x1B[0m \x1B[90m 78 |\x1B[39m\x1B[0m',
      '\x1B[0m\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 79 |\x1B[39m \x1B[36mimport\x1B[39m \x1B[32m"path"\x1B[39m\x1B[33m;\x1B[39m\x1B[0m',
      '\x1B[0m \x1B[90m    |\x1B[39m         \x1B[31m\x1B[1m^\x1B[22m\x1B[39m\x1B[0m',
      '\x1B[0m \x1B[90m 80 |\x1B[39m\x1B[0m',
    ].join('\n');
    expect(stripAnsi(stripMetroInfo(input))).toMatchInlineSnapshot(`
      "  77 | }
        78 |
      > 79 | import "path";
           |         ^
        80 |"
    `);
  });
});

describe(isNodeStdLibraryModule, () => {
  it(`returns true for node standard library modules`, () => {
    for (const moduleName of ['node:fs', 'fs/promises', 'net']) {
      expect(isNodeStdLibraryModule(moduleName)).toBe(true);
    }
  });
  it(`returns false for etc modules`, () => {
    for (const moduleName of ['expo', '@expo/config', 'uuid']) {
      expect(isNodeStdLibraryModule(moduleName)).toBe(false);
    }
  });
});

describe(formatUsingNodeStandardLibraryError, () => {
  it(`formats node standard library error`, () => {
    const format = formatUsingNodeStandardLibraryError('/Users/evanbacon/my-app', {
      message: 'foobar',
      originModulePath: '/Users/evanbacon/my-app/App.js',
      targetModuleName: 'path',
    } as any);
    expect(stripAnsi(format)).toMatchInlineSnapshot(`
      "You attempted attempted to import the Node standard library module "path" from "App.js".
      It failed because the native React runtime does not include the Node standard library.
      Learn more: https://docs.expo.dev/workflow/using-libraries/#using-third-party-libraries"
    `);
  });
});
