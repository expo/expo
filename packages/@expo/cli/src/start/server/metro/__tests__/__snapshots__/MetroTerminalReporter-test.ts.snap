// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_getBundleStatusMessage should format standard progress for a server invocation 1`] = `"λ ./index.js ▓▓▓▓▓▓▓▓░░░░░░░░ 50.0% ( 50/100)"`;
